From 360bcbcbe61131f800b33cc910c38ad14193399d Mon Sep 17 00:00:00 2001
From: David Andrey <david.andrey@netmodule.com>
Date: Fri, 7 Dec 2012 16:46:09 +0100
Subject: [PATCH 04/20]  ARM: xilinx: U-Boot udelay < 1000 FIX

Rework the __udelay function of U-Boot Zynq Arch to handle
delay < 1000 usec

Signed-off-by:David Andrey david.andrey@netmodule.com
---
 arch/arm/cpu/armv7/zynq/timer.c |   38 ++++++++++++++++++++++++++++++++------
 1 files changed, 32 insertions(+), 6 deletions(-)

diff --git a/arch/arm/cpu/armv7/zynq/timer.c b/arch/arm/cpu/armv7/zynq/timer.c
index 323e7b5..9e6894b 100644
--- a/arch/arm/cpu/armv7/zynq/timer.c
+++ b/arch/arm/cpu/armv7/zynq/timer.c
@@ -114,15 +114,41 @@ ulong get_timer_masked(void)
 
 void __udelay(unsigned long usec)
 {
-	unsigned long long tmp;
-	ulong tmo;
+	u32 countTicks;
+	u32 timeEnd;
+	u32 timeDiff;
+	u32 timeNow;
 
-	tmo = usec / (1000000 / CONFIG_SYS_HZ);
-	tmp = get_ticks() + tmo; /* Get current timestamp */
+	if (usec == 0)
+		return;
 
-	while (get_ticks() < tmp) { /* Loop till event */
-		 /* NOP */;
+	countTicks = (u32) (((unsigned long long) TIMER_TICK_HZ * usec) / 1000000);
+
+	timeEnd = readl(&timer_base->counter) - countTicks; /* decrementing timer */
+
+#if TIMER_LOAD_VAL != 0xFFFFFFFF
+	/* do not manage multiple overflow*/
+	if (countTicks >= TIMER_LOAD_VAL) {
+		countTicks = TIMER_LOAD_VAL -1;
 	}
+#endif
+
+	do {
+		timeNow = readl(&timer_base->counter);
+
+		if (timeNow >= timeEnd) {
+			/* normal case */
+			timeDiff = timeNow - timeEnd;
+		} else {
+			if ((TIMER_LOAD_VAL - timeEnd + timeNow) <= countTicks) {
+				/* overflow */
+				timeDiff = TIMER_LOAD_VAL - timeEnd + timeNow;
+			} else {
+				/* missed the exact match */
+				break;
+			}
+		}
+	} while (timeDiff > 0);
 }
 
 /* Timer without interrupts */
-- 
1.7.4.1

